pkg/pillar/cmd/zedrouter/probe.go

```diff
@@ -0,0 +1,534 @@
+	}
+	iteration++
+}
+
+func probeCheckStatus(status *types.NetworkInstanceStatus) {
+	if len(status.PInfo) == 0 {
+		return
+	}
+
+	f := make([]bool, 3) // non-valid, free, non-free
+	f[intfTypeFree] = true
+	f[intfTypeNonFree] = false
+	// check probe stats from Free pool first
+	// continue to the non-Free if there is no available good link in the Free pool
+	for c := intfTypeFree; c <= intfTypeNonFree; c++ {
+		var numOfUps int
+		probeCheckStatusUseType(status, f[c])
+		currIntf := status.CurrentUplinkIntf
+		if currIntf != "" {
+			if currinfo, ok := status.PInfo[currIntf]; ok {
+				numOfUps = infoUpCount(currinfo)
+			}
+		}
+		if numOfUps > 0 {
+			break
+		}
+	}
+	log.Infof("probeCheckStatus: %s current Uplink Intf %s, prev %s\n", status.BridgeName, status.CurrentUplinkIntf, status.PrevUplinkIntf)
+}
+
+// How to determine the time to switch to another interface
+// -- compare only within the same interface class: ether, lte, sat
+// -- Random assign one intf intially
+// -- each intf has 3 types of states: both local and remote report UP, only one is UP, both are Down
+// -- try to pick and switch to the one has the highest degree of UPs
+// -- otherwise, don't switch
+func probeCheckStatusUseType(status *types.NetworkInstanceStatus, free bool) {
+	var numOfUps, upCnt int
+	currIntf := status.CurrentUplinkIntf
+	prevIntf := currIntf
+	if currIntf == "" {
+		for _, info := range status.PInfo {
+			if free != info.IsFree {
+				continue
+			}
+			upCnt = infoUpCount(info)
+			if currIntf == "" { // assign any intf for now
+				currIntf = info.IfName
+				numOfUps = upCnt
+				continue
+			} else {
+				if numOfUps < upCnt {
+					currIntf = info.IfName
+					numOfUps = upCnt
+				}
+			}
+		}
+	} else {
+		if _, ok := status.PInfo[currIntf]; !ok {
+			// should not happen, zero it out, come next time to process
+			log.Errorf("probeCheckStatusUseType: current Uplink Intf %s error", currIntf)
+			status.CurrentUplinkIntf = ""
+			return
+		}
+		currinfo := status.PInfo[currIntf]
+		numOfUps = infoUpCount(currinfo)
+		if free == currinfo.IsFree && numOfUps == 2 { // good, no need to change
+			return
+		}
+		if free != currinfo.IsFree { // from a lower intf class, start over again
+			currIntf = ""
+		}
+		for _, info := range status.PInfo {
+			if free != info.IsFree {
+				continue
+			}
+			if strings.Compare(info.IfName, currIntf) == 0 {
+				continue
+			}
+			upCnt = infoUpCount(info)
+			if numOfUps < upCnt {
+				currIntf = info.IfName
+				numOfUps = upCnt
+			}
+		}
+	}
+	if strings.Compare(status.CurrentUplinkIntf, currIntf) != 0 {
+		log.Infof("probeCheckStatusUseType: changing from %s to %s\n",
+			status.CurrentUplinkIntf, currIntf)
+		status.CurrentUplinkIntf = currIntf
+		status.PrevUplinkIntf = prevIntf
+		status.NeedIntfUpdate = true
+	}
+}
+
+func infoUpCount(info types.ProbeInfo) int {
+	var upCnt int
+	if info.GatewayUP && info.RemoteHostUP {
+		upCnt = 2
+	} else if info.GatewayUP || info.RemoteHostUP {
```

Comment:
Are these really equivalent? Depends what probe targets the user configured. Hence I don't know how to compare a GatewayUP and a RemoteHostUP in general.

Reply:
right now we treat them the same weight. i think they eventually will converge, only temporarily it will have things like one link local is good, the other link only remote is good, assume both local and remote used to work fine.